app/assets/javascripts/discourse/app/lib/hashtag-autocomplete.js

```diff
@@ -62,7 +58,65 @@ export function hashtagTriggerRule(textarea, opts) {
   return true;
 }
 
-function _setupExperimental(context, $textArea, siteSettings, afterComplete) {
+const checkedHashtags = new Set();
+let seenHashtags = {};
+
+// NOTE: For future maintainers, the hashtag lookup here does not take
+// into account mixed contexts -- for instance, a chat quote inside a post
+// or a post quote inside a chat message, so this may
+// not provide an accurate priority lookup for hashtags without a ::type suffix in those
+// cases.
+export function fetchUnseenHashtagsInContext(orderedContextTypes, slugs) {
+  return ajax("/hashtags", {
+    data: { slugs, order: orderedContextTypes },
+  }).then((response) => {
+    Object.keys(response).forEach((type) => {
+      seenHashtags[type] = seenHashtags[type] || {};
+      response[type].forEach((item) => {
+        seenHashtags[type][item.ref] = seenHashtags[type][item.ref] || item;
+      });
+    });
+    slugs.forEach(checkedHashtags.add, checkedHashtags);
+  });
+}
+
+export function linkSeenHashtagsInContext(orderedContextTypes, elem) {
+  const hashtagSpans = [...(elem?.querySelectorAll("span.hashtag-raw") || [])];
+  if (hashtagSpans.length === 0) {
+    return [];
+  }
+  const slugs = [...hashtagSpans.map((hashtag) => hashtag.innerText)];
+
+  hashtagSpans.forEach((hashtagSpan, index) => {
+    let slug = slugs[index];
+
+    orderedContextTypes.forEach((type) => {
+      // remove type suffixes
+      const typePostfix = `::${type}`;
+      if (slug.endsWith(typePostfix)) {
+        slug = slug.slice(0, slug.length - typePostfix.length);
+      }
+
+      // replace raw span for the hashtag with a cooked one
+      const matchingSeenHashtag = seenHashtags[type]?.[slug];
+      if (matchingSeenHashtag) {
+        _replaceSeenHashtagPlaceholder(type, hashtagSpan, matchingSeenHashtag);
+      }
+    });
```

Comment:
wonder if all of this shouldn't be in `_replaceSeenHashtagPlaceholder` or maybe another function

Reply:
Yes fair enough, will tweak this